### PR TITLE
plat-stm32mp1: move to rstctrl framework

### PR DESCRIPTION
**Pull request by etienne-lms**

### CODE REVIEW

File: core/drivers/rstctrl/stm32_rstctrl.c [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+DEFINE_DT_DRIVER(stm32_rstctrl_dt_driver) = {
+	.name = "stm32_rstctrl",
+	.type = DT_DRIVER_RSTCTRL,
+	.match_table = stm32_rstctrl_match_table,
+	.probe = stm32_rstctrl_provider_probe,
+};
+#endif /*CFG_EMBED_DTB*/
+
+/* stm32mp1 legacy reset control API function */
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/rstctrl/stm32_rstctrl.c:227]: Well... legacy functions `stm32_reset_assert()`, `stm32_reset_deassert()` and `stm32_reset_assert_deassert_mcu()` could be removed, yes, once the 2nd patch in this series is applied.

We will still need `stm32mp_rcc_reset_id_to_rstctrl()` for drivers that do not probe from the DT (hence cannot use DT means to get the expected reset controller instances). An example is the SCMI service, 2nd commit of this series, that calls.

I can add a 3rd commit in this series to remove these old stm32mp APIs.

File: core/arch/arm/plat-stm32mp1/conf.mk [hunk 2]
```diff
@@ -98,6 +99,10 @@ ifeq ($(call cfg-one-enabled, CFG_STM32_CRYP),y)
 $(call force,CFG_STM32_CRYPTO_DRIVER,y)
 endif
 
+ifeq ($(CFG_STM32_RSTCTRL),y)
+CFG_DRIVERS_RSTCTRL ?= y
+endif
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/conf.mk:104]: > Is this STM32 reset controller driver still useable when CFG_DRIVERS_RSTCTRL=n

The intend is to move to the new API, so yes old `stm32_reset_*()` API functions could be removed, see next comment.

```
CFG_DRIVERS_RSTCTRL ?= $(CFG_STM32_RSTCTRL)
$(eval $(call cfg-depends-all,CFG_STM32_RSTCTRL,CFG_DRIVERS_RSTCTRL))
```
looks ok to me.

File: core/drivers/rstctrl/stm32_rstctrl.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jforissier) [reply core/drivers/rstctrl/stm32_rstctrl.c:227]: > I can add a 3rd commit in this series to remove these old stm32mp APIs.

That would be good IMO.

File: core/arch/arm/plat-stm32mp1/conf.mk (hunk 2 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/conf.mk]: comment addressed straight in 1st [rebased commit](https://github.com/OP-TEE/optee_os/pull/5119/commits/a2152d9e51ba6a661bfb9d428ba9c87264327a1f#diff-f250837c324370445a1a2b29ff2b798e93ef5814f07097eb2e85d12aea5b0b99R102-R103).

File: core/drivers/rstctrl/stm32_rstctrl.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/rstctrl/stm32_rstctrl.c]: done

**Comment by etienne-lms** [comment]
Rebased (with _plat-stm32mp1/conf.mk_ fixed in 1st commit) and appended 2 more commits:
- change stm32_cryp to drop legacy API functions
- remove stm32mp1 reset control legacy API functions

**Review by jforissier — COMMENTED** [review]
"even **in** case of driver probe deferral error **as** stm32_cryp is not yet defined as a DT_DRIVER."

`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by etienne-lms** [comment]
Fixed date in new source file.
Fixed spelling in commit log.
Applied A-b tag.

**Pull request merged**